### PR TITLE
Write nested entity reference link(s) in request/response

### PR DESCRIPTION
**Pull request by xuzhg**
### Issues

* Write nested entity reference link(s) in request/response

### Description

* Write nested entity reference link(s) in request/response

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs [hunk 1]
```diff
@@ -1003,6 +1003,17 @@ protected override void EndNestedResourceInfoWithContent(ODataNestedResourceInfo
...
+            {
+                JsonLightNestedResourceInfoScope navigationLinkScope = (JsonLightNestedResourceInfoScope)this.CurrentScope;
+
+                // If we wrote entity reference links for a collection navigation property but no
+                // resource set afterwards, we have to now close the array of links.
+                if (navigationLinkScope.EntityReferenceLinkWritten && !navigationLinkScope.ResourceSetWritten && nestedResourceInfo.IsCollection.Value)
+                {
+                    this.jsonWriter.EndArrayScope();
+                }
+            }
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: Why don't we just move this before line 987, and remove 989-997 (and this else condition)?

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs [hunk 2]
```diff
@@ -1027,8 +1037,17 @@ protected override void WriteEntityReferenceInNavigationLinkContent(ODataNestedR
 
             if (!navigationLinkScope.EntityReferenceLinkWritten)
             {
-                // Write the property annotation for the entity reference link(s)
-                this.odataAnnotationWriter.WritePropertyAnnotationName(parentNestedResourceInfo.Name, ODataAnnotationNames.ODataBind);
+                if (!this.jsonLightOutputContext.WritingResponse &&
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: >!this.jsonLightOutputContext.WritingResponse && [](start = 20, length = 47)

why does this depend upon whether or not we're writing a response?

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs [hunk 3]
```diff
@@ -1038,8 +1057,35 @@ protected override void WriteEntityReferenceInNavigationLinkContent(ODataNestedR
                 navigationLinkScope.EntityReferenceLinkWritten = true;
             }
 
-            Debug.Assert(entityReferenceLink.Url != null, "The entity reference link Url should have been validated by now.");
-            this.jsonWriter.WriteValue(this.jsonLightResourceSerializer.UriToString(entityReferenceLink.Url));
+            if (!this.jsonLightOutputContext.WritingResponse &&
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: >!this.jsonLightOutputContext.WritingResponse [](start = 16, length = 44)

why is this dependent upon request/response?

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs [hunk 4]
```diff
@@ -1038,8 +1057,35 @@ protected override void WriteEntityReferenceInNavigationLinkContent(ODataNestedR
                 navigationLinkScope.EntityReferenceLinkWritten = true;
             }
 
-            Debug.Assert(entityReferenceLink.Url != null, "The entity reference link Url should have been validated by now.");
-            this.jsonWriter.WriteValue(this.jsonLightResourceSerializer.UriToString(entityReferenceLink.Url));
+            if (!this.jsonLightOutputContext.WritingResponse &&
+                (this.Version == null || this.Version < ODataVersion.V401))
+            {
+                Debug.Assert(entityReferenceLink.Url != null, "The entity reference link Url should have been validated by now.");
+                this.jsonWriter.WriteValue(this.jsonLightResourceSerializer.UriToString(entityReferenceLink.Url));
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: should this assert be moved out of the conditional? don't we assume the entityReferenceLink.Url is set regardless of format?

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs (hunk 1 above)
Comment (xuzhg) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: Good. That's more clear. #Resolved.

---
In reply to: [279949030](https://github.com/OData/odata.net/pull/1459#discussion_r279949030) [](ancestors = 279949030)

File: test/FunctionalTests/Microsoft.OData.Core.Tests/IntegrationTests/Evaluation/AutoComputePayloadMetadataInJsonIntegrationTests.cs [hunk 5]
```diff
@@ -752,6 +752,210 @@ public void WritingDynamicComplexPropertyWithModelSpecifiedInFullMetadataMode_40
...
+        private string GetEntityReferenceLinksPayload(ODataVersion version, bool request, bool collection)
+        {
+            MemoryStream stream = new MemoryStream();
+            ODataWriter odataWriter = GetODataWriterForEntityReferenceLink(version, request, stream);
+
+            odataWriter.WriteStart(new ODataResource { Properties = new[] { new ODataProperty { Name = "CustomerId", Value = 7 } } });
+            if (collection)
+            {
+                odataWriter.WriteStart(new ODataNestedResourceInfo { Name = "BillOrders", IsCollection = true });
+                odataWriter.WriteEntityReferenceLink(new ODataEntityReferenceLink { Url = new Uri("http://svc/Orders(2)") });
```
Comment (mikepizzo) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/IntegrationTests/Evaluation/AutoComputePayloadMetadataInJsonIntegrationTests.cs]: >Url [](start = 84, length = 3)

it's too bad the property is named "Url", and not "Id", but not something to change at this point. Maybe in 8.0.

Comment (xuzhg) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/IntegrationTests/Evaluation/AutoComputePayloadMetadataInJsonIntegrationTests.cs]: Can't agree more.

**Pull request merged**